rex/utilities/bc_utils.py

```diff
@@ -0,0 +1,294 @@
+        """
+
+        msg = f'params must be 2D array but received {type(params)}'
+        assert isinstance(params, np.ndarray), msg
+
+        if len(params.shape) == 1:
+            params = np.expand_dims(params, 0)
+
+        msg = (f'params must be 2D array of shape ({arr_shape[1]}, N) '
+               f'but received shape {params.shape}')
+        assert len(params.shape) == 2, msg
+        assert params.shape[0] == arr_shape[1], msg
+
+        if self.scipy_dist is not None:
+            params = [params[:, i] for i in range(params.shape[1])]
+
+        return params
+
+    def _get_quantiles(self, n_samples):
+        """If dist='empirical', this will get the quantile values for the CDF
+        x-values specified in the input params"""
+
+        if self.sampling == 'linear':
+            quantiles = sample_q_linear(n_samples)
+        elif self.sampling == 'log':
+            quantiles = sample_q_log(n_samples, self.log_base)
+        elif self.sampling == 'invlog':
+            quantiles = sample_q_invlog(n_samples, self.log_base)
+        else:
+            msg = ('sampling option must be linear, log, or invlog, but '
+                   'received: {}'.format(self.sampling))
+            logger.error(msg)
+            raise KeyError(msg)
+
+        return quantiles
+
+    def cdf(self, x, params):
+        """Run the CDF function e.g., convert physical variable to quantile"""
+
+        if self.scipy_dist is None:
+            p = np.zeros_like(x)
+            for idx in range(x.shape[1]):
+                xp = params[idx, :]
+                fp = self._get_quantiles(len(xp))
+                p[:, idx] = np.interp(x[:, idx], xp, fp)
+        else:
+            p = self.scipy_dist.cdf(x, *params)
+
+        return p
+
+    def ppf(self, p, params):
+        """Run the inverse CDF function (percent point function) e.g., convert
+        quantile to physical variable"""
+
+        if self.scipy_dist is None:
+            x = np.zeros_like(p)
+            for idx in range(p.shape[1]):
+                fp = params[idx, :]
+                xp = self._get_quantiles(len(fp))
+                x[:, idx] = np.interp(p[:, idx], xp, fp)
+        else:
+            x = self.scipy_dist.ppf(p, *params)
+
+        return x
+
+    def __call__(self, arr):
+        """Run the QDM function to bias correct an array
+
+        Parameters
+        ----------
+        arr : np.ndarray
+            2D array of values in shape (time, space)
+
+        Returns
+        -------
+        arr : np.ndarray
+            Bias corrected copy of the input array with same shape.
+        """
+
+        if len(arr.shape) == 1:
+            arr = np.expand_dims(arr, 1)
+
+        params_oh = self._clean_params(self.params_oh, arr.shape)
+        params_mh = self._clean_params(self.params_mh, arr.shape)
+        params_mf = self._clean_params(self.params_mf, arr.shape)
+
+        q_mf = self.cdf(arr, params_mf)  # Tau_m_p
```

Comment:
if these comment refer to equation terms could you add the full equation in comments as well?

Reply:
Yeah gotcha. Will do but i'll also add the equation numbers. I'm not sure the equation terms are any more readable than the code. For example:

<img width="303" alt="image" src="https://github.com/NREL/rex/assets/12700356/afceea56-7e22-458d-ba09-0481cf82d3d5">

vs. 

`x_oh = self.ppf(q_mf, params_oh)  # x^_o:m_h:p`

Reply:
But is this a term in a larger equation or no? I was thinking they were terms and that a complete equation would provide context but maybe thats not true.

Reply:
There's only one code line that is just a part of a larger equation from cannon. I added more comments, see the new commit. I think it's as good as we can get.